rmf_task/include/rmf_task/State.hpp

```diff
@@ -0,0 +1,111 @@
+ *
+*/
+
+#ifndef RMF_TASK__STATE_HPP
+#define RMF_TASK__STATE_HPP
+
+#include <chrono>
+
+#include <rmf_utils/impl_ptr.hpp>
+
+#include <rmf_traffic/Time.hpp>
+#include <rmf_traffic/agv/Planner.hpp>
+
+#include <rmf_task/CompositeData.hpp>
+
+namespace rmf_task {
+
+//==============================================================================
+class State : public CompositeData
+{
+public:
+
+  /// The current waypoint of the robot state
+  RMF_TASK_DEFINE_COMPONENT(std::size_t, CurrentWaypoint);
+  std::optional<std::size_t> waypoint() const;
+  State& waypoint(std::size_t new_waypoint);
+
+  /// The current orientation of the robot state
+  RMF_TASK_DEFINE_COMPONENT(double, CurrentOrientation);
+  std::optional<double> orientation() const;
+  State& orientation(double new_orientation);
+
+  /// The current time for the robot
+  RMF_TASK_DEFINE_COMPONENT(rmf_traffic::Time, CurrentTime);
+  std::optional<rmf_traffic::Time> time() const;
+  State& time(rmf_traffic::Time new_time);
+
+  /// The dedicated charging point for this robot
+  // TODO(MXG): Consider removing this field and using some kind of
+  // ChargingStrategy abstract interface to determine where and how the robots
+  // should be charging.
+  RMF_TASK_DEFINE_COMPONENT(std::size_t, DedicatedChargingPoint);
+  std::optional<std::size_t> dedicated_charging_waypoint() const;
+  State& dedicated_charging_waypoint(std::size_t new_charging_waypoint);
+
+  /// The current battery state of charge of the robot. This value is between
+  /// 0.0 and 1.0.
+  RMF_TASK_DEFINE_COMPONENT(double, CurrentBatterySoC);
+  std::optional<double> battery_soc() const;
+  State& battery_soc(double new_battery_soc);
+
+  /// Load the basic state components expected for the planner.
+  ///
+  /// \param[in] location
+  ///   The robot's initial location data.
+  ///
+  /// \param[in] charging_point
+  ///   The robot's dedicated charging point.
+  ///
+  /// \param[in] battery_soc
+  ///   The robot's initial battery state of charge.
+  State& load_basic(
+    const rmf_traffic::agv::Plan::Start& location,
+    std::size_t charging_point,
+    double battery_soc);
+
+  /// Load the plan start into the State. The location info will be split into
+  /// CurrentWaypoint, CurrentOrientation, and CurrentTime data.
+  State& load(const rmf_traffic::agv::Plan::Start& location);
+
+  /// Project an rmf_traffic::agv::Plan::Start from this State.
+  ///
+  /// If CurrentWaypoint is unavailable, this will return a std::nullopt. For
+  /// any other components that are unavailable (CurrentOrientation or
+  /// CurrentTime), the given default values will be used.
+  ///
+  /// \param[in] default_orientation
+  ///   The orientation value that will be used if CurrentOrientation is not
+  ///   available in this State.
+  ///
+  /// \param[in] default_time
+  ///   The time value that will be used if default_time is not available in
+  ///   this State.
+  std::optional<rmf_traffic::agv::Plan::Start> project_plan_start(
```

Comment:
`project_` and `extract_` have different implications, as mentioned in the documentation.

`project_plan_start()` only requires the `CurrentWaypoint` component to be available. Everything else will get filled in with default values. This should be used if you're handling a **partial** state estimate and aren't concerned about details like time or orientation.

`extract_plan_start()` requires all relevant components to be available, or else it will return a `std::nullopt`. This should be used if you believe you're handling a **fully defined** state estimate and want to generate a plan that starts from that state.

They can't be overloaded with the same name because they can both be called with the same number of arguments (zero). We could overload them if we remove the default arguments from `project_plan_start`, but I'd like to keep the distinction between `project_` and `extract_` since they're meant to be used in different contexts.

Reply:
Thanks for explaining that distinction.